### DISCUSSION TITLE
StyleTTS2 Training from Scratch Notebooks

### DISCUSSION THREAD

**Discussion by martinambrus** [yl4579/StyleTTS2#144] [Show and tell]
I'm currently learning how to train a custom StyleTTS2 model from scratch.

I'm very new to this and thanks to this amazing project and its community, I've already gained a considerable amount of knowledge. Here, I'd like to share that knowledge with you.

To that end, I created 2 [Jupyter Notebooks](https://jupyter.org/) that I use for my own model training and audio samples preparation.

I'd like to stress out that ***these are only my own methods and findings*** and there probably is a better way to do things, especially in the audio preparation part. But since I tried - and failed - to find a reliable automated method there, some manual fine-tuning steps are still required in order to perfect the audio input.

My 2 Notebooks:
- [StyleTTS2 Model Training from Scratch](https://colab.research.google.com/drive/1PdcCcTUUYAxBM3_mDT2_UNz0ZoUkMee_?usp=sharing)
- [WAV files preparation Notebook](https://colab.research.google.com/drive/1ZK-2lAV2DokrN92sYPJRl47XyuuluCJe?usp=sharing)

These notebooks can be used on [Google Colab](https://colab.research.google.com/) but also outside of it, on a dedicated cloud machine running Jupyter.

~~Perhaps I should clarify that ***I'm still in the process of learning*** and as such, I'm yet to create a production-ready model.~~
See [my last update below](https://github.com/yl4579/StyleTTS2/discussions/144#discussioncomment-8066717) for more information on a production-ready model training and results.

These Notebooks previously served me to create a ***low-quality model from ~150 WAV files*** with a duration of 1 - 2.5 seconds as a proof of concept. At the present time, I use them to finalize my production-ready model training.
To that end, I used 2 [TensorDock Cloud GPU Machines](https://www.tensordock.com/):
- for the ***1st training phase***, a beefy one (costing approx. $3.38/hr) with:
  - 8xRTX A6000 48GB GPUs
  - 120GB RAM
  - 20 CPUs
  - 500GB of HDD space

  I've done ***1000 epochs***, from which only the first ***400*** were needed - at least from what I can see in the validation loss info (around 0.38 - 0.4). Those 400 epochs were done very quickly (in less then 1 or 2 hours, if I recall correctly) on this small data set.
- for the ***2nd training phase***, I used a similar machine but with only a single GPU, since there is still a bug in code that doesn't allow us to use DDP (accelerate) training in this phase. The cost went down to approx. $0.65/hr and it took about 2-4 hours to finish ***100 epochs*** on this small data set.

I'm currently working towards the creation of a large high-quality corpus, spanning approx. 45 hours of audio.
The source for this corpus are 4 audio-books read by another high-quality (now decomissioned) TTS voice to which I had a commercial license.
My goal is to try and train a similarly-sounding voice model by utilizing StyleTTS2.
Here is an example of how the original TTS voice sounds like: https://jmp.sh/zOrGoel3

I should also mention that I already used the set of those ~150 WAV files to fine-tune StyleTTS2 to the new voice, and even with as little data as those files provide, I was able to achieve a very good voice transfer quality.

I hope these Notebooks will help someone to automate their training, too :)

**Comment by platform-kit** [comment]
Thank you for this! Can't wait to try it. How much of a corpus did you need to train from scratch (minutes/hours of audio) and how long did it take to get a production quality result, when training from scratch? (# of epochs? hours? And on what hardware?)

**Reply by martinambrus** [reply]
I edited my original text to provide more info.

**Reply by platform-kit** [reply]
Thanks. BTW your audio sample link has expired.

**Reply by martinambrus** [reply]
sorry, didn't know they expire on that platform... link updated to something more stable

**Comment by martinambrus** [comment]
I thought that I'd shed some light on the **progress so far**.

It's been a bit slow, especially since I was preparing all of the WAV files and training / validation text by hand to minimize error rate. I currently have around **10 hours of audio** ready in **8534 WAV files**, spanning from 1 to 22 seconds.

At the end, I decided to give it a try with what I have, since the weeks cutting and syncing WAVS to text were really getting boring. I used [VAST.ai](https://cloud.vast.ai) for GPU rental and got the **first stage** trained up to **200 epochs** in about **3,5 hours**. The hardware and settings I used for this training were as follows:

- **_machine_**
  - 112 GB HDD
  - 86 GB RAM
  - 8x A100X 80GB GPUs
  - AMD EPYC 7302 16-Core Processor, 64 cores total
- **_config_**
  - ```epochs_1st``` = 200
  - ```batch_size``` = 408 (until epoch _50_, since that's where additional validation steps kick in), then 288 _after epoch 50_
  - ```max_len``` = 1760 (for the longest 22 seconds audio file -> _22 / 0.0125 = 1760_ )
  - ```save_freq``` = 10 (to save HDD space, I only saved every 10th checkpoint)
- **_cost_**
  - $11.231/h x 3,5 hours = **_$33.7_**

The **second stage** is still in progress, currently at **epoch 67**. It's taking considerably longer because it can only run on a single GPU ( due to bug #7 ). Therefore, I'm running it **every day for about 12-18 hours** which brought the budget up to almost **$200** by now. I decided to run it only when I have control over the process, since I didn't know how many epochs I can use that the A100 memory can handle.

At the end, I used the following parameters for training:
- **_machine_**
  - 112 GB HDD
  - 86 GB RAM
  - 1x A100X 80GB GPU
  - AMD EPYC 7302 16-Core Processor, 64 cores total
- **_config_**
  - ```epochs_2nd``` = 200
  - ```batch_size``` = 32 (until epoch _20_, since that's where additional validation steps kick in), then I had to tune this down from 14 gradually to only 6 (which is stable but really slow now)
  - ```max_len``` = 1760 (for the longest 22 seconds audio file -> _22 / 0.0125 = 1760_ )
  - ```save_freq``` = 1 (because it takes ages for a checkpoint and I don't want to loose even a single one + this allows me to test each checkpoint as it's created)
- **_cost (so far)_**
  - $1.863/h x 108 hours = **_$201.2_**

Example of how inference sounds at epoch 65 as compared to the original TTS voice: https://jmp.sh/aWMQe69G
Original TTS: https://jmp.sh/zOrGoel3

**Reply by casic** [reply]
And may be you work only for English ???

**Reply by martinambrus** [reply]
Yes, I'm training an English model. In my experiences, I found that multilingual models tend to sometimes switch into a different language mid-sentence and then their pronounciation and the whole inference is wrong. I prefer working with single-language models at the moment.

**Reply by effusiveperiscope** [reply]
Neat! Do you mind sharing loss graphs from your training? I haven't seen very many examples of successful full training

**Reply by martinambrus** [reply]
I'm afraid I don't have that data from stage 1 training and I don't know how to make a graph from a log, so I'm attaching the log I have for phase 2.
[train.log](https://github.com/yl4579/StyleTTS2/files/13897648/train.log)

**Reply by effusiveperiscope** [reply]
Thanks. Have you tried using tensorboard?

**Reply by martinambrus** [reply]
Good question. I've never used it before. But the second VM is deleted as well now and all I have is this log for the time being, so I'm afraid I can't help with those graphs at this point. I guess if you needed graphs, they'd need to be re-created from the data from the log.

**Reply by vishalsantoshi** [reply]
Were you able to complete the training ?

**Reply by martinambrus** [reply]
not yet, I needed to redirect my finances to something else for the time being, so I'm still at epoch 75 of the 2nd stage

**Reply by tracertea** [reply]
@martinambrus do you have an example of the original audio you trained on to compare the style transfer?

**Reply by martinambrus** [reply]
@tracertea see the two links at the end of the original comment

**Reply by tracertea** [reply]
Thanks @martinambrus I miss read the original post mentioning the original source was a TTS.

**Reply by ethan-digi** [reply]
@martinambrus Would you be willing to share your loss curves/logs for second stage training? I'm also re-training from scratch and fear something is amiss with diffusion. Would be greatly appreciated if you can.

**Reply by martinambrus** [reply]
hi @ethan-digi - I've been digging through what I backed up from the training machine and unfortunately, I can't find data from 2nd stage training... I'm sorry, I'd gladly send them to you but I'm afraid I did not back those up :(

**Reply by ethan-digi** [reply]
@martinambrus All good, I appreciate you checking regardless. Let me ask one thing then - do you recall, vaguely, what the loss curve/pattern looked like for diffusion? Mine isn't meaningfully decreasing, but other losses are

**Reply by martinambrus** [reply]
@ethan-digi I don't recall exactly but I don't think I saw anything off while watching the log - all of the losses decreased as expected... but perhaps you could open an issue in the repo with your output, so other people who also trained from scratch may be able to help you out?

**Reply by martinambrus** [reply]
If anyone would be interested in seeing some curves and train data, check out the link in this issue: https://github.com/yl4579/StyleTTS2/issues/281

**Comment by junylee11** [comment]
I couldn't increase the batch size per a100 by more than 8 when training. How did you manage to advance learning with such a large batch size?

**Reply by martinambrus** [reply]
This is going to very much depend on the lengths of your audio files. If they are long, you won't be able to go too high, since the data would not fit into a single A100's memory.

Also, I was using a100 with 80GB of VRAM. They are not as common as their 40GB counterparts, so make sure you use the bigger ones for better performance.

**Reply by junylee11** [reply]
For me, it's about 4 to 6 seconds per audio.
How many seconds does your audio average per unit?

I have a batch size of 8 per gpus and 4 in stage 2.
Increasing it further causes a cuda memory error.

I'm amazed that your batch size is so big.

**Reply by martinambrus** [reply]
I've got about 8-11 seconds per audio file average, some are even 15 or 20 seconds long. So your data should fit in there easily. Do you use 80GB VRAM A100 units?

**Reply by junylee11** [reply]
Yes, I used a100, 80gb gpu. (max_len : 800)
I couldn't raise the batch size to more than 8 due to the vram issue in stage 2 learning.
However, you said that you conducted the second stage of training with the batch size of 32.

I wonder why there is such a difference.

**Reply by martinambrus** [reply]
Did you run out of memory even before epoch 20 during stage 2? Because after that point, even I had to lower the batch size to 4 with my data set. But before that point, I started at 32 and was lowering it as the model started to run out of memory, continuing from the last checkpoint.

**Reply by junylee11** [reply]
Yes, I was having this problem at joint epoch.
Now I understand! Thank you.

I'd like to ask you one more question.
I am currently in the process of learning 2stage joint epoch with about 100 hours of data.
I'm learning using 8 units of a100 and it takes over 20 hours per epoch.

Will I get satisfactory performance with just 1 or 2 epoch?
Is there any way to reduce my learning time?

(Batch size and max_len are set to optimal values.)

**Reply by martinambrus** [reply]
The thing is stage 2 currently doesn't work using more than 1 unit, since parallel processing still has an unresolved bug (issue #7). So unfortunatelly, we're left with slow 2nd stage learning for the time being. Decreasing the number of epochs won't train a satisfactory model, since it needs to train with probably even more than 200 epochs to be perfect. With 2 epochs, you wouldn't get anywhere near a HQ result.

**Reply by junylee11** [reply]
Thank you for your answer.

In the case of the libri data model of @yl4579, we produced a satisfactory performance model with only 20 epoch.
Of course, the performance will be lower than the result of going up to 200epoch, but I think we can manage to make the result.

There is no other alternative than DDP learning, so I have no choice but to wait for a long training. :(

**Reply by martinambrus** [reply]
Thanks. But did you do a fine-tuning on the existing @yl4579's libriTTS model in those 20 epochs or did you train from scratch?

I only trained a single language model, so I can't quite comment on multi-lang one. However, 20 epochs sound like a very low number if I'm training from scratch. At least for my own EN model, it required at least 75 epochs of stage 2 training for the model to stop having various artifacts in the infered output.

**Reply by junylee11** [reply]
Thank you for your answer.
When I referred to the paper, it seems that I trained from the beginning.

I just came up with a solution to my problem.
So how about reducing the amount of data in 2stage learning?
We reduce the 100 hours of the dataset used for stage 1 learning to 20 hours to conduct stage 2 learning.

No matter how hard I think about it, it's too wasteful of resources and time to learn with the amount of the first dataset.

Will reducing the amount of data in the middle of learning cause fatal flaws in learning?

**Reply by Jedi92614** [reply]
Hi martinambrus
Would you be able to train and release Vietnamese model for Stylett2. I would be able to help pay for all cost of machine time.
Best

**Reply by martinambrus** [reply]
Hi @Jedi92614, since I don't speak Vietnamese, it wouldn't be feasible for me to [train a PL-BERT model](https://github.com/yl4579/PL-BERT) that's be needed for this language. Also, I wouldn't be able to prepare training data correctly for the same reason. If you'd have the training WAV files and PL-BERT model trained for Vietnamese, I can help you with training the model but I'd need to have those resources ready first.

**Reply by Jedi92614** [reply]
Hey martinambrus,

Thanks a lot for the quick response! I've gone through your training instructions carefully, and I've got to say, you've put together a really well-thought-out guide. It's great to see the effort you've invested to help others with their model training. Thanks for taking the time to do that!

I'm reaching out because I'm interested in having you as a consultant for a Vietnamese StyleTTS project. We're assembling a team to handle the groundwork, and we'd love to have your expert eyes on our progress. The plan is to share our outputs with you at each critical stage for your review and guidance on whether to move forward or make revisions. The project is open source and aimed at supporting the Vietnamese community.

To get this collaboration going, we'll need to set up a way to communicate directly. If you're interested, could you please get in touch with me at Michael@Atumjstudio.com? Your expertise would be incredibly valuable, and I'm excited about the possibility of working together.

Thanks again for all your help!

Best,
Michael

**Reply by martinambrus** [reply]
@Jedi92614 that sounds interesting. Also, thank you for your kind words. I'm sending you an e-mail.

**Reply by martinambrus** [reply]
@Jedi92614 not sure whether you're still working on the project but there's some nice insight on training a new PL-BERT (if needed) as well as the advantages of training your language-specific ASR as well in issue #257

**Reply by martinambrus** [reply]
@junylee11 I've resumed my attempts on training the English model and came to a conclusion that I was lucky before, since probably most of the training data were sub-10-second samples, so I could advance quite fast. Right now I'm re-training the model again from scratch (I also found a couple of bugs in the training notebook that I created). This time, I can't use values as high since my training and validation set contain longer sentences which don't fit into the A100 memory as neatly as those previous samples did. I think I also need to update the splitting script, so it splits on WAV durations more than on randomness.

**Reply by Lwasinam** [reply]
@martinambrus 
I'm currently training on 500hrs of data
Second stage and I get OOM errors after a while. I'm using 8xA100s 40GB
My batch size is 16 (2 for each GPU) this is the least I can go 

My Max_len is about 250, increasing it any further would definitely use more memory usage and my not accommodate joint epoch training

**Reply by martinambrus** [reply]
@Lwasinam from my experience, I couldn't fit more than a batch of 12 into a single A100 80GB, so you're probably using a much bigger batch than what the GPU can process. Also, setting `max_len` to 250 will only ever process 3.125s of each audio file. That's probably not what you want, unless your audio files are at max 3.125s long. If they are longer, I would suggest using a higher `max_len` value (ideally to the maximum length of your longest audio file). Otherwise, your model will learn to use randomly and abruptly-cut audio endings, which will not help the final quality. Also, you will most probably need to lower your batch size and/or trying to use 80GB GPUs instead of 40GB ones. One last thing - you don't need 500hrs of data if they are from a single speaker. For a single speaker, around 10 hours of data is quite enough - even less if you can afford to spend more time training (i.e. going through more epochs).

**Reply by Lwasinam** [reply]
Thanks so much

**Reply by Lwasinam** [reply]
@martinambrus when training on lets say a 10hr single speaker dataset, 
during second stage training, are the logged outputs clear from the first epoch, I mean clearly understandable?

**Reply by martinambrus** [reply]
I'm not sure what you mean but here are the data if you want to review them yourself: https://github.com/yl4579/StyleTTS2/issues/281

**Comment by DevOps920719** [comment]
hello. @martinambrus , Nice to meet you. I have just saw your old posts here. 
Could we co-operate together? My email is priyam.jha0124@gmail.com
I am really exited about the possibility to co-operate with you.
Thank you.

**Reply by martinambrus** [reply]
hi, what sort of cooperation you're after?

**Reply by moesaeed** [reply]
@martinambrus Could we do a paid collaboration on training for different languages?
We have 400 hours chucks into 8-15 sec with transcripts+ we have the required GPU.

**Reply by martinambrus** [reply]
@moesaeed hi - so far I did English and a test for Czech language but even with 10 hours of audio, I wasn't able to reproduce the quality of the original English TTS. I'm open for discussion though. Feel free to contact me at 09wmd42of@mozmail.com
